examples/all-clusters-app/linux/diagnostic-logs-provider-delegate-impl.cpp

```diff
@@ -0,0 +1,181 @@
+                                           const chip::Optional<uint64_t> & outTimeSinceBoot)
+{
+    VerifyOrReturnValue(IsValidIntent(intent), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto filePath = GetFilePathForIntent(intent);
+    VerifyOrReturnValue(filePath.HasValue(), CHIP_ERROR_NOT_FOUND);
+
+    auto fp = fopen(filePath.Value().c_str(), "rb");
+    VerifyOrReturnValue(nullptr != fp && errno != ENOENT, CHIP_ERROR_NOT_FOUND);
+    VerifyOrReturnValue(nullptr != fp, CHIP_ERROR_INTERNAL);
+
+    mLogSessionHandle++;
+    // If the session handle rolls over to UINT16_MAX which is invalid, reset to 0.
+    VerifyOrDo(mLogSessionHandle != kInvalidLogSessionHandle, mLogSessionHandle = 0);
+
+    outHandle                 = mLogSessionHandle;
+    mFiles[mLogSessionHandle] = fp;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LogProvider::EndLogCollection(LogSessionHandle sessionHandle)
+{
+    VerifyOrReturnValue(sessionHandle != kInvalidLogSessionHandle, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnValue(mFiles.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto fp = mFiles[sessionHandle];
+    fclose(fp);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LogProvider::CollectLog(LogSessionHandle sessionHandle, MutableByteSpan & outBuffer, bool & outIsEndOfLog)
+{
+    VerifyOrReturnValue(sessionHandle != kInvalidLogSessionHandle, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnValue(mFiles.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto fp       = mFiles[sessionHandle];
+    auto fileSize = GetFileSize(fp);
+    auto count    = std::min(fileSize, outBuffer.size());
+
+    if (!chip::CanCastTo<off_t>(count))
+    {
+        outBuffer.reduce_size(0);
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    auto bytesRead = fread(outBuffer.data(), 1, count, fp);
+    if (!chip::CanCastTo<size_t>(bytesRead))
+    {
+        outBuffer.reduce_size(0);
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    outBuffer.reduce_size(static_cast<size_t>(bytesRead));
+    outIsEndOfLog = fileSize == static_cast<size_t>(ftell(fp));
+    return CHIP_NO_ERROR;
+}
+
+size_t LogProvider::GetSizeForIntent(IntentEnum intent)
+{
+    VerifyOrReturnValue(IsValidIntent(intent), 0);
+
+    auto filePath = GetFilePathForIntent(intent);
+    VerifyOrReturnValue(filePath.HasValue(), 0);
+
+    auto fp = fopen(filePath.Value().c_str(), "rb");
+    VerifyOrReturnValue(nullptr != fp, 0);
+
+    return GetFileSize(fp);
+}
+
+chip::Optional<std::string> LogProvider::GetFilePathForIntent(IntentEnum intent)
+{
+    chip::Optional<std::string> rv = NullOptional;
+
+    switch (intent)
+    {
+    case IntentEnum::kEndUserSupport:
+        rv = mEndUserSupportLogFilePath;
+        break;
+    case IntentEnum::kNetworkDiag:
+        rv = mNetworkDiagnosticsLogFilePath;
+        break;
+    case IntentEnum::kCrashLogs:
+        rv = mCrashLogFilePath;
+        break;
+    case IntentEnum::kUnknownEnumValue:
```

Comment:
how about using `default` instead? that way it does not look like one could return null_optional.